refapp/SKWebApi/Skills/ChatSkill.cs

```diff
@@ -0,0 +1,285 @@
+            text: formattedMessage,
+            id: currentTime,
+            cancel: context.CancellationToken
+        );
+    }
+
+    private async IAsyncEnumerable<MemoryQueryResult> GetAllMemoriesAsync(SKContext context)
+    {
+        var allCollections = await context.Memory.GetCollectionsAsync(context.CancellationToken);
+        var allChatMessageCollections = allCollections.Where(collection => collection != "ChatMessages");
+        IList<MemoryQueryResult> allChatMessageMemories = new List<MemoryQueryResult>();
+        try
+        {
+            foreach (var collection in allChatMessageCollections)
+            {
+                var results = await context.Memory.SearchAsync(
+                    collection,
+                    "abc",                  // dummy query
+                    limit: 1,
+                    minRelevanceScore: 0.0, // no relevance required since the collection only has one entry
+                    cancel: context.CancellationToken
+                ).ToListAsync();
+                allChatMessageMemories.Add(results.First());
+            }
+        }
+        catch (Exception e)
+        {
+            Console.WriteLine("Exception while retrieving memories: {0}", e.Message);
+            throw e;
+        }
+
+        foreach (var memory in allChatMessageMemories.OrderBy(memory => memory.Id))
+        {
+            yield return memory;
+        }
+    }
+
+    private async Task<MemoryQueryResult> GetLatestMemoryAsync(SKContext context)
+    {
+        var allMemories = this.GetAllMemoriesAsync(context);
+        return await allMemories.FirstAsync();
+    }
+
+    private CompleteRequestSettings GetChatResponseCompletionSettings()
+    {
+        var completionSettings = new CompleteRequestSettings();
+        completionSettings.MaxTokens = SystemPromptDefaults.ResponseTokenLimit;
+        completionSettings.Temperature = SystemPromptDefaults.ResponseTemperature;
+        completionSettings.TopP = SystemPromptDefaults.ResponseTopP;
+        completionSettings.FrequencyPenalty = SystemPromptDefaults.ResponseFrequencyPenalty;
+        completionSettings.PresencePenalty = SystemPromptDefaults.ResponsePresencePenalty;
+        return completionSettings;
+    }
+
+    private CompleteRequestSettings GetIntentCompletionSettings()
+    {
+        var completionSettings = new CompleteRequestSettings();
+        completionSettings.MaxTokens = SystemPromptDefaults.ResponseTokenLimit;
+        completionSettings.Temperature = SystemPromptDefaults.IntentTemperature;
+        completionSettings.TopP = SystemPromptDefaults.IntentTopP;
+        completionSettings.FrequencyPenalty = SystemPromptDefaults.IntentFrequencyPenalty;
+        completionSettings.PresencePenalty = SystemPromptDefaults.IntentPresencePenalty;
+        completionSettings.StopSequences = new string[] { "] bot:" };
+        return completionSettings;
+    }
+
+    private int EstimateTokenCount(string text)
```

Comment:
We've a token calculator that just came in from Devis. At least add a TODO here to use it.